### PR TITLE
[filbeat][azure-blob-storage] - Adding support for Microsoft Entra ID RBAC authentication

### LINKED ISSUE #40434: [filebeat][azure-blob-storage] - Add support for authorization via Microsoft Entra ID / RBAC
**Describe the enhancement:**  Add support for authorization via Microsoft Entra ID / RBAC 

**Describe a specific use case for the enhancement or feature:** Currently the azure-blob-storage input handles client auth via shared key credentials and connection strings, but this is inherently insecure according to recent recommendations by Microsoft as outlined by this [article](https://learn.microsoft.com/en-us/azure/storage/common/shared-key-authorization-prevent?tabs=portal). To mitigate security issues we need to add auth support via Microsoft Entra ID.  This will require us to overhaul the client auth process and change certain internal auth structs to get it working.  The sample code below shows an example of how to get Microsoft Entra ID auth working with the **blob storage sdk utilizing the azidentity sdk**. 

```go
 // Define the Azure AD credentials
    tenantID := "your-tenant-id"
    clientID := "your-client-id"
    clientSecret := "your-client-secret"

    // Create a ClientSecretCredential
    cred, err := azidentity.NewClientSecretCredential(tenantID, clientID, clientSecret, nil)
    if err != nil {
        log.Fatalf("failed to create client secret credential: %v", err)
    }

    // Define the Azure Blob Storage account URL
    accountName := "your_account_name"
    blobURL := fmt.Sprintf("https://%s.blob.core.windows.net/", accountName)

    // Create a new BlobServiceClient with the Azure credential
    serviceClient, err := azblob.NewClient(blobURL, cred, nil)
    if err != nil {
        log.Fatalf("failed to create blob service client: %v", err)
    }
```
Using this process of authentication we will require the **tenantID, clientID & clientSecret** as inputs from the users end. More  info regarding this is detailed [here](https://github.com/Azure/azure-sdk-for-go/tree/main/sdk/storage/azblob#readme) and [here](https://learn.microsoft.com/en-us/azure/storage/blobs/authorize-access-azure-active-directory).

This process however creates a service client of a different type ```("github.com/Azure/azure-sdk-for-go/sdk/storage/azblob").Client``` and is incompatible with our current service client which is of type ```("github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/service").Client```, so the necessary changes need to be made under the hood to get this working.

### PR DESCRIPTION
**Pull request by ShourieG**
## Type of change
- Enhancement 
- Docs

## Proposed commit message
Added support for Microsoft Entra ID RBAC authentication.
Added mock tests by injecting the gock transport layer in the azure client.
Added some config tests with the new config options.

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Disruptive User Impact
No Impact

## Live Testing
Live testing was performed using our internal Azure dev environment.
**Process followed:** 
1. Fresh storage account and container created and then uploaded with dummy files.
2. New service principle app created.
3. blobOwner role granted to service app via Entra ID IAM configs.
4. client_id, tenant_id and client_secret of the service app used to poll poll the blob storage.
5. Records were output to a local file with success.

## Author's Checklist

- [ ]

## Related issues
- Closes #40434

## Testing
We performed live testing internally using an active azure account and the new authentication system is working properly after assigning the service principal app with the **blobOwner** permission level.

## Screenshots
Screenshot 2024-09-19 at 6 29 07 PM
Screenshot 2024-09-19 at 6 30 46 PM

### CODE REVIEW

File: x-pack/filebeat/input/azureblobstorage/auth_test.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/h2non/gock"
```
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/auth_test.go:15]: Do we need this? From what I can see, an implementation of `http.RoundTripper` would satisfy the testing requirements that tests here add, arguably more cleanly.

File: x-pack/filebeat/input/azureblobstorage/client.go [hunk 2]
```diff
@@ -88,6 +112,17 @@ func fetchBlobClientWithConnectionString(connectionString string, containerName
 	return blobClient, nil
 }
 
+func fetchBlobClientWithOAuth2(url string, credential *azidentity.ClientSecretCredential, oauth2Cfg *OAuth2Config) (*blob.Client, error) {
+	blobClient, err := blob.NewClient(url, credential, &blob.ClientOptions{
+		ClientOptions: oauth2Cfg.clientOptions,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Error fetching blob client for url : %s, error : %w", url, err)
```
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/client.go:120]:
```suggestion
		return nil, fmt.Errorf("failed to fetch blob client for %s: %w", url, err)
```

File: x-pack/filebeat/input/azureblobstorage/config.go [hunk 3]
```diff
@@ -44,19 +47,35 @@ type fileSelectorConfig struct {
...
+
 func defaultConfig() config {
 	return config{
 		AccountName: "some_account",
 	}
 }
+
+func (c config) Validate() error {
+	if c.Auth.OAuth2 != nil && (c.Auth.OAuth2.ClientID == "" || c.Auth.OAuth2.ClientSecret == "" || c.Auth.OAuth2.TenantID == "") {
+		return fmt.Errorf("client_id, client_secret and tenant_id are required for OAuth2 auth")
```
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/config.go:78]:
```suggestion
		return errors.New("client_id, client_secret and tenant_id are required for OAuth2 auth")
```

File: x-pack/filebeat/input/azureblobstorage/auth_test.go (hunk 1 above)
Comment (andrewkroh) [reply x-pack/filebeat/input/azureblobstorage/auth_test.go:15]: I had the same thought when I gave this a quick scan yesterday. The declarative setup is nice, but I'm not sure it's worth another dependency.

Comment (ShourieG) [reply x-pack/filebeat/input/azureblobstorage/auth_test.go:15]: If we go that route then we would lose the context to the serv.URL thats local to each test. serv.URL would be required to set & override the urls in the response JSON without which the sdk will always try to call 
```https://login.microsoftonline.com``` for the token validation check, as this is hardwired into the sdk and cannot be overridden by conventional means.

Comment (efd6) [reply x-pack/filebeat/input/azureblobstorage/auth_test.go:15]: > If we go that route then we would lose the context to the serv.URL thats local to each test.

Why? There is nothing in the gock code that could not be done easily and more transparently with local code. AFAICS the mock code provides author convenience at the cost of transparency for the maintainer.

Comment (ShourieG) [reply x-pack/filebeat/input/azureblobstorage/auth_test.go]: @efd6, have removed gock and simplified the test.

File: x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc [hunk 4]
```diff
@@ -148,6 +149,37 @@ Now let's explore the configuration attributes a bit more elaborately.
...
 
+[id="attrib-auth-oauth2"]
+[float]
+==== `auth.oauth2`
+
+This attribute contains the Microsoft Entra ID RBAC authentication credentials for a secure connection to the Azure Blob Storage. The `auth.oauth2` attribute contains the following sub-attributes:
+
+    1. `client_id` : The client ID of the Azure Entra ID application.
+    2. `client_secret` : The client secret of the Azure Entra ID application.
+    3. `tenant_id` : The tenant ID of the Azure Entra ID application.
```
Comment (efd6) [inline x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc:160]:
```suggestion
    1. `client_id`: The client ID of the Azure Entra ID application.
    2. `client_secret`: The client secret of the Azure Entra ID application.
    3. `tenant_id`: The tenant ID of the Azure Entra ID application.
```

File: x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc [hunk 5]
```diff
@@ -148,6 +149,37 @@ Now let's explore the configuration attributes a bit more elaborately.
...
+    tenant_id: "some_tenant_id"
+  containers:
+  - name: container_1
+    max_workers: 3
+    poll: true
+    poll_interval: 10s
+----
+How to setup the `auth.oauth2` credentials can be found in the Azure documentation https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app[here]
+
+NOTE: According to our internal testing it seems that we require at least an access level of **blobOwner** for the service principle to be able to read the blobs. If you are facing any issues with the access level, please ensure that the access level is set to **blobOwner**.
```
Comment (efd6) [inline x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc:181]:
```suggestion
NOTE: According to our internal testing it seems that we require at least an access level of **blobOwner** for the service principle to be able to read the blobs. If you are facing any issues with the access level, ensure that the access level is set to **blobOwner**.
```

**Review by efd6 — APPROVED** [review]
nits only

**Pull request merged**